tests/functional/test_csv_viewer.py

```diff
@@ -99,12 +111,162 @@ def test_csv_renders_all_text(live_server, browser, csv_file):
+        first_row_sorted_asc = rows[0]
+
+    # sort ascending by nth column
+    sort_button_1.click()
+
+    # wait for sorting to finish
+    expect(
+        page.locator(
+            f"thead th:nth-child({column_sort_index + 2}) .icon.datatable-icon--ascending"
+        )
+    ).to_be_visible()
+
+    # check first row is as expected
+    for item in first_row_sorted_asc:
+        expect(page.locator("tbody tr:nth-child(1)").first).to_contain_text(
+            item, timeout=1
+        )
+
+    # sort descending by nth column
+    sort_button_1.click()
+
+    # wait for sorting to finish
+    expect(
+        page.locator(
+            f"thead th:nth-child({column_sort_index + 2}) .icon.datatable-icon--descending"
+        )
+    ).to_be_visible()
+
+    # Check cell in sorted column first row is as expected.
+    # The sort behaviour in python/javascript is sufficiently different that
+    # it's hard to get this correct after the first sort. So rather than adding
+    # unnecessary complexity to the test, we just check that the first value in
+    # the sorted column is as expected
+    expect(
+        page.locator("tbody tr:nth-child(1)").first.locator(
+            f"td:nth-child({column_sort_index + 2})"
+        )
+    ).to_contain_text(last_row_sorted_column_value, timeout=1)
+
+    # sort ascending by (n+1)th column
+    sort_button_2.click()
+
+    # wait for sorting to finish
+    expect(
+        page.locator(
+            f"thead th:nth-child({column_sort_index + 3}) .icon.datatable-icon--ascending"
+        )
+    ).to_be_visible()
+
+    expect(
+        page.locator("tbody tr:nth-child(1)").first.locator(
+            f"td:nth-child({column_sort_index + 3})"
+        )
+    ).to_contain_text(first_row_second_sort_column_value, timeout=1)
+
+    # See note at the top of file
     if "coverage" not in sys.modules.keys():  # pragma: no cover
         context.close()
+
+
+# Ensure every element in a large csv table is eventually visible with scolling
+def test_csv_scroll(live_server, page, context):
+    workspace = factories.create_workspace("my-workspace")
+    num_rows = 1000
+
+    factories.write_workspace_file(
+        workspace,
+        "outputs/file1.csv",
+        "\r\n".join([f"value,value,value{i}" for i in range(num_rows)]),
+    )
+    login_as_user(
+        live_server,
+        context,
+        user_dict=factories.create_api_user(
+            username="author",
+            workspaces={
+                "my-workspace": factories.create_api_workspace(project="Project 2"),
+            },
+        ),
+    )
+
+    page.goto(
+        live_server.url + workspace.get_contents_url(UrlPath("outputs/file1.csv"))
+    )
+
+    page.locator("tbody").hover()
+    for i in range(num_rows):
+        row_locator = page.get_by_text(f"value{i}", exact=True)
+        if not row_locator.is_visible():  # pragma: no cover (if we ever revert to non-virtualized tables then this is never True)
```

Comment:
Could this use `pragma: no branch` instead? I think that would mean that it'd pass coverage while this `if` is always True, but fail if it's never True (in which case, we'd probably want to know, so we can fix the test)

Reply:
Ah I didn't know that - thanks for the tip.